source/generators/ECECGenerator.cc

```diff
@@ -0,0 +1,157 @@
+// parameters the atomic number, mass number and energy level of the
+// daughter isotope of interest.
+//
+// The NEXT Collaboration
+// ----------------------------------------------------------------------------
+
+
+#include "ECECGenerator.h"
+#include "GeometryBase.h"
+#include "DetectorConstruction.h"
+#include "FactoryBase.h"
+
+#include <G4GenericMessenger.hh>
+#include <G4RunManager.hh>
+#include <G4Event.hh>
+#include <G4AtomicShell.hh>
+#include <G4AtomicShellEnumerator.hh>
+#include <G4ParticleDefinition.hh>
+#include <G4PrimaryVertex.hh>
+#include <G4PrimaryParticle.hh>
+#include <G4DynamicParticle.hh>
+#include <G4UAtomicDeexcitation.hh>
+
+using namespace nexus;
+
+REGISTER_CLASS(ECECGenerator, G4VPrimaryGenerator)
+
+ECECGenerator::ECECGenerator():
+  G4VPrimaryGenerator(),
+  atomic_number_(-1),
+  shell_name_("K"), // K, L1, L2, L3, M1, ..., M5
+  region_("ACTIVE"),
+  msg_  (nullptr),
+  geom_ (nullptr),
+  shell_(nullptr),
+  atom_ (nullptr)
+{
+  msg_ = new G4GenericMessenger(this, "/Generator/ECEC/",
+                                "Control commands of the ECEC primary generator.");
+
+  G4GenericMessenger::Command& atomic_number_cmd =
+    msg_->DeclareProperty("atomic_number", atomic_number_,
+                          "Atomic number of the ion.");
+  atomic_number_cmd.SetParameterName("atomic_number", false);
+  atomic_number_cmd.SetRange("atomic_number > 0");
+
+  msg_->DeclareProperty("shell", shell_name_,
+                        "Shell from which the electron is captured.");
+
+  msg_->DeclareProperty("region", region_,
+                        "Region of the geometry where vertices will be generated.");
+
+}
+
+
+ECECGenerator::~ECECGenerator()
+{
+  delete msg_;
+  delete shell_;
+  delete atom_;
+}
+
+
+G4AtomicShellEnumerator ECECGenerator::GetShellID(G4String shell)
+{
+       if (shell == "K" ) return fKShell;
+  else if (shell == "L1") return fL1Shell;
+  else if (shell == "L2") return fL2Shell;
+  else if (shell == "L3") return fL3Shell;
+  else if (shell == "M1") return fM1Shell;
+  else if (shell == "M2") return fM2Shell;
+  else if (shell == "M3") return fM3Shell;
+  else if (shell == "M4") return fM4Shell;
+  else if (shell == "M5") return fM5Shell;
+  else
+    G4Exception("[ECECGenerator]", "GetShellID()", FatalException,
+                "Cannot recognize shell.");
+
+  return fKShell; // Doesn't mean anything. Just to avoid -Wreturn-type
+}
+
+
+void ECECGenerator::Initialize() {
+  const DetectorConstruction* detconst =
+    dynamic_cast<const DetectorConstruction*>
+    (G4RunManager::GetRunManager()->GetUserDetectorConstruction());
+
+  if (!detconst)
+    G4Exception("[ECECGenerator]", "ECECGenerator()", FatalException,
+                "Unable to load geometry.");
+
+  geom_ = detconst->GetGeometry();
+
+  atom_ = new G4UAtomicDeexcitation();
+  atom_->SetFluo (true);
+  atom_->SetAuger(true);
+  atom_->SetCutForSecondaryPhotons(0);
```

Comment:
I think this was just the first thing I tried and then I realized that I needed to set it to >0 in the GenerateParticles function. At the time I didn't know that GenerateParticles overwrites these values. So probably these two lines can even be removed. What do you prefer?

Reply:
For what I can see in the code, those thresholds are used exclusively in the `GenerateParticles()` function, therefore it should be exactly the same; on top of that, that function requires to be passed those arguments. The whole structure looks quite confusing to me, but it's Geant4... I would remove the setters at the beginning, since they don't do anything.

Reply:
As discussed offline, these setters have no effect as the `GenerateParticle` method overrides them.